src/metrics/mongoose_metrics_mam_hooks.erl

```diff
@@ -16,135 +16,158 @@
+        {mam_set_prefs, HostType, fun ?MODULE:mam_set_prefs/3, #{}, 50},
+        {mam_get_prefs, HostType, fun ?MODULE:mam_get_prefs/3, #{}, 50},
+        {mam_archive_message, HostType, fun ?MODULE:mam_archive_message/3, #{}, 50},
+        {mam_remove_archive, HostType, fun ?MODULE:mam_remove_archive/3, #{}, 50},
+        {mam_lookup_messages, HostType, fun ?MODULE:mam_lookup_messages/3, #{}, 100},
+        {mam_flush_messages, HostType, fun ?MODULE:mam_flush_messages/3, #{}, 50}
     ].
 
 %% @doc Here will be declared which hooks should be registered when mod_mam_muc is enabled.
--spec get_mam_muc_hooks(_) -> [ejabberd_hooks:hook(), ...].
-get_mam_muc_hooks(Host) ->
+-spec get_mam_muc_hooks(_) -> gen_hook:hook_list().
+get_mam_muc_hooks(HostType) ->
     [
-        {mam_muc_set_prefs, Host, ?MODULE, mam_muc_set_prefs, 50},
-        {mam_muc_get_prefs, Host, ?MODULE, mam_muc_get_prefs, 50},
-        {mam_muc_archive_message, Host, ?MODULE, mam_muc_archive_message, 50},
-        {mam_muc_remove_archive, Host, ?MODULE, mam_muc_remove_archive, 50},
-        {mam_muc_lookup_messages, Host, ?MODULE, mam_muc_lookup_messages, 100},
-        {mam_muc_flush_messages, Host, ?MODULE, mam_muc_flush_messages, 50}
+        {mam_muc_set_prefs, HostType, fun ?MODULE:mam_muc_set_prefs/3, #{}, 50},
+        {mam_muc_get_prefs, HostType, fun ?MODULE:mam_muc_get_prefs/3, #{}, 50},
+        {mam_muc_archive_message, HostType, fun ?MODULE:mam_muc_archive_message/3, #{}, 50},
+        {mam_muc_remove_archive, HostType, fun ?MODULE:mam_muc_remove_archive/3, #{}, 50},
+        {mam_muc_lookup_messages, HostType, fun ?MODULE:mam_muc_lookup_messages/3, #{}, 100},
+        {mam_muc_flush_messages, HostType, fun ?MODULE:mam_muc_flush_messages/3, #{}, 50}
     ].
 
--spec mam_get_prefs(Result :: any(),
-                    Host :: jid:server(),
-                    _ArcID :: mod_mam:archive_id(),
-                    _ArcJID :: jid:jid()) -> any().
-mam_get_prefs(Result, Host, _ArcID, _ArcJID) ->
-    mongoose_metrics:update(Host, modMamPrefsGets, 1),
-    Result.
-
--spec mam_set_prefs(Result :: any(), Host :: jid:server(),
-    _ArcID :: mod_mam:archive_id(), _ArcJID :: jid:jid(),
-    _DefaultMode :: any(), _AlwaysJIDs :: [jid:literal_jid()],
-    _NeverJIDs :: [jid:literal_jid()]) -> any().
-mam_set_prefs(Result, Host, _ArcID, _ArcJID, _DefaultMode, _AlwaysJIDs, _NeverJIDs) ->
-    mongoose_metrics:update(Host, modMamPrefsSets, 1),
-    Result.
-
--spec mam_remove_archive(Acc :: map(),
-                         Host :: jid:server(),
-                         _ArcID :: mod_mam:archive_id(),
-                         _ArcJID :: jid:jid()) -> metrics_notify_return().
-mam_remove_archive(Acc, Host, _ArcID, _ArcJID) ->
-    mongoose_metrics:update(Host, modMamArchiveRemoved, 1),
-    Acc.
-
+-spec mam_get_prefs(Acc, Params, Extra) -> {ok, Acc} when
+    Acc :: any(),
+    Params :: map(),
+    Extra :: #{host_type := mongooseim:host_type()}.
```

Comment:
I think that all these declaring a map with the host_type are a bit repetitive. It should perhaps be an exported type from `gen_hook`. For example, see how I did it in the c2s feature branch here: https://github.com/esl/MongooseIM/blob/49eb66ceb5bb6df8a0d8883916c42a2524cdd232/src/gen_hook.erl#L35 Then hooks can use that exported type and reduce duplication. Feel free to copy-paste those, I'll deal with the merges 👌🏽

Reply:
done 👍